htsget-http/src/middleware/auth.rs

```diff
@@ -0,0 +1,593 @@
+      ));
+    };
+
+    self.fetch_from_url(&query_url.to_string()).await
+  }
+
+  /// Validate the restrictions, returning an error if the user is not authorized.
+  pub fn validate_restrictions(
+    restrictions: AuthorizationRestrictions,
+    request: Request,
+    endpoint: Endpoint,
+  ) -> HtsGetResult<()> {
+    // Find all rules matching the path.
+    let matching_rules = restrictions
+      .htsget_auth()
+      .iter()
+      .filter(|rule| {
+        // If this path is a direct match then just return that.
+        if rule.path().strip_prefix("/").unwrap_or(rule.path())
+          == request.path().strip_prefix("/").unwrap_or(request.path())
+        {
+          return true;
+        }
+
+        // Otherwise, try and parse it as a regex.
+        Regex::new(rule.path()).is_ok_and(|regex| regex.is_match(request.path()))
+      })
+      .collect::<Vec<_>>();
+
+    // If any of the rules allow all reference names (nothing set in the rule) then the user is authorized.
+    let None = matching_rules
+      .iter()
+      .find(|rule| rule.reference_names().is_none())
+    else {
+      return Ok(());
+    };
+
+    let format = match_format_from_query(&endpoint, request.query())?;
+    let query = convert_to_query(request, format)?;
+    let matching_restriction = matching_rules
+      .iter()
+      .flat_map(|rule| rule.reference_names().unwrap_or_default())
+      .find(|restriction| {
+        // The reference name should match exactly.
+        let name_match = Some(restriction.name()) == query.reference_name();
+        // The format should match if it's defined, otherwise it allows any format.
+        let format_match =
+          restriction.format().is_none() || restriction.format() == Some(query.format());
+        // The interval should match exactly, considering undefined start or end ranges.
+        let interval_match = restriction.interval().contains_interval(query.interval());
+
+        name_match && format_match && interval_match
+      });
+
+    // If the matching rule with the restriction was found, then the user is authorized, otherwise
+    // it is a permission denied response.
+    if matching_restriction.is_some() {
+      Ok(())
+    } else {
+      Err(HtsGetError::PermissionDenied(
+        "failed to authorize user based on authorization service restrictions".to_string(),
+      ))
+    }
+  }
+
+  /// Validate the authorization flow, returning an error if the user is not authorized.
+  /// This performs the following steps:
+  ///
+  /// 1. Finds the JWT decoding key from the config or by querying a JWKS url.
+  /// 2. Validates the JWT token according to the config.
+  /// 3. Queries the authorization service for restrictions based on the config or JWT claims.
+  /// 4. Validates the restrictions to determine if the user is authorized.
+  pub async fn validate_authorization(
+    &self,
+    request: Request,
+    endpoint: Endpoint,
+  ) -> HtsGetResult<()> {
```

Comment:
This is the main logic that handles the authorization. The rest of the functions in this module do the authorization logic too. A lot of the other changes are just hooking it up to the server and config parsing.